common/benchmark_utils.h

```diff
@@ -0,0 +1,355 @@
+		assert(info.hdr);
+	} else {
+		fprintf(stderr, "hdr is already set.\n");
+		return;
+	}
+
+	while(fgets(buf, 4096, info.fp)) {
+		if (strrchr(buf, '\n')) { /* end of first line */
+			memcpy(info.hdr + (cnt - 1) * 4096, buf, strlen(buf));
+			break;
+		} else {
+			info.hdr_len = ++cnt * 4096;
+			info.hdr = realloc(info.hdr, info.hdr_len);
+			memcpy(info.hdr + (cnt - 1) * 4096, buf, 4096);
+		}
+	}
+
+	__set_cols();
+}
+
+#if 0
+/**
+ * Print the data to the given file specified by info.fname.
+ */
+static inline void __print_table(double data)
+{
+	static unsigned int col;
+
+	if (!info.hdr_printed) {
+		char *tmp = strdup(info.hdr), *ptr, *cur;
+
+
+		int i = 0;
+
+		for (cur = tmp; i < info.cols - 1; i++) {
+			ptr = strchr(cur, ',');
+
+			if (ptr)
+				*ptr = 0;
+
+			fprintf(stdout, "%-*s", FIELD_WIDTH, cur);
+
+			cur += strlen(cur) + 1;
+		}
+
+		fprintf(stdout, "%s.\n", cur);
+
+		free(tmp);
+		info.hdr_printed = 1;
+	}
+
+	if (info.date_len && col == info.date_col) {
+		time_t t;
+		struct tm *lt;
+
+		time(&t);
+		lt = localtime(&t);
+
+		if (lt) {
+			fprintf(stdout, info.date_str, lt->tm_year,
+				lt->tm_mon, lt->tm_mday, lt->tm_hour, lt->tm_min,
+				lt->tm_sec);
+			fprintf(stdout, ".");
+		}
+		col = (col + 1) % info.cols;
+	}
+
+	fprintf(stdout, col != (info.cols - 1) ? "%*.*lf" : "%*.*lf\n",
+		FIELD_WIDTH, FLOAT_PRECISION, data);
+
+	fflush(stdout);
+
+	col = (col + 1) % info.cols;
+}
+#endif
+
+/**
+ * Print the data to the given file specified by info.fname.
+ */
+static inline void __print_csv(double data)
+{
+	static unsigned int col;
+
+	/* If the csv file is empty, fill in the header. */
+	if (info.stats.st_size == 0) {
+		fprintf(info.fp, "%s\n", info.hdr);
+		info.stats.st_size = strlen(info.hdr);
+	}
+
+	/* Print the date to the appropriate column. */
+	if (~info.date_col && col == info.date_col) {
+		time_t t;
+		struct tm *lt;
+
+		time(&t);
+		lt = localtime(&t);
+
+		if (lt) {
+			fprintf(info.fp, info.date_str, lt->tm_year % 100,
+				lt->tm_mon, lt->tm_mday, lt->tm_hour, lt->tm_min,
+				lt->tm_sec);
+		}
+	}
+
+	fprintf(info.fp, col != (info.cols - 1) ? "%.*lf," : ".*lf\n",
+		FLOAT_PRECISION, data);
+
+	col = (col + 1) % info.cols;
+}
+
+/*******************************************************************************
+ * Exposed utility fns
+ ******************************************************************************/
+/**
+ * This function must be called before print_data() or fini_info().
+ * Initialize the info structure, hdr or path must be non-NULL.
+ * Note: The file pointed to by path must already exist for the
+ * csv file to be written to.
+ *
+ * @assumptions The path must be non-NULL and valid, the header from an
+ * existing csv file pointed to by path will always be used in place of the
+ * hdr param
+ *
+ * @param hdr  the table header.
+ * @param path the path to the csv file.
+ * @return 0 on success, -1 on error.
+ */
+static inline int init_info(char *hdr, char *path)
```

Comment:
I recommend duplicating the header string rather than storing it. You don't know and can't assume they won't allocate this string on the stack somewhere.